### DISCUSSION TITLE
How to prevent `TypeError` onChange of an input field?

### DISCUSSION THREAD

**Discussion by obliviga** [react-hook-form/react-hook-form#9236] [Q&A] [answered]
Hello, I'm using the [Adobe Spectrum React](https://react-spectrum.adobe.com/react-spectrum/TextField.html) library with `react-hook-form`. I have two input fields that I want to validate, but I get a `TypeError` every time the `onChange` event fires: `TypeError: Cannot read properties of undefined (reading 'name')`.

I created a CodeSandbox to demonstrate this issue here: https://codesandbox.io/s/react-hook-form-with-adobe-react-spectrum-mb9chl?file=/src/App.js

Any ideas on why an error is being thrown here?
<img width="424" alt="image" src="https://user-images.githubusercontent.com/5934582/196251422-0c68946d-f6ab-464e-abaa-c09a9d260871.png">

**Comment by bluebill1049** [comment]
I don't think that component support uncontrolled method with register, you will have to use controller for it.

**Comment by Moshyfawn** [comment, marked as answer]
React Spectrum _(RS)_ `onChange` returns the result value of the user input and RHF expects it's event instead as the `register` API was designed to be "DOM-centered". It's the nature of RS library that dictates for them to not expose the DOM event, as they've built a platform agnostic system that there might not be a DOM element on some platforms.

Please, refer to [this RS discussion](https://github.com/adobe/react-spectrum/issues/1860#issuecomment-849833808) for more details

**Reply by Moshyfawn** [reply]
That said, @bluebill1049 is there a reason the `register` API doesn't support passing the user input result to `onChange`, only the DOM event? Do you think it's a good idea for me to try supporting it?

**Reply by bluebill1049** [reply]
Thanks, @Moshyfawn we actually had that type of logic in v6 and v5. This creates noise in the evaluation of the actual field value by introducing more logic than is needed, and in v7 I have stripped it back to only support native input reference only. However, is there are strong demands outside we can consider to reintroude the logic again.

**Reply by bluebill1049** [reply]
In addition, this will not work correctly with the focus management as well, so probably it's better to use `useController` instead so you can wire up the input reference.

**Reply by Moshyfawn** [reply]
I see, @bluebill1049; valid points!

I don't see an immediate need to allow this behaviour as RHF provides the controller API that covers most of the cases where external input API makes it difficult to wire it up to RHF. I mostly just didn't know where you stood on this, hence the question ;) it's pretty common for an input to expose its DOM event object with the exception of a few and most of them have a valid reason to do so
